extractor/docker/docker_test.go

```diff
@@ -338,3 +396,209 @@ func TestDockerExtractor_Extract(t *testing.T) {
+			requiredFiles: []string{"somerandomfilethatdoesntexistinthelayer"},
+			cacheWB:       false,
+		},
+	}
+
+	for _, tc := range testCases {
+		inputDigest := digest.Digest("sha256:62d8908bee94c202b2d35224a221aaa2058318bfa9879fa541efaecba272331b")
+		layerData, err := ioutil.ReadFile("../../utils/testdata/testdir.tar.gz")
+		assert.NoError(t, err)
+
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			httpPath := r.URL.String()
+			switch {
+			case strings.Contains(httpPath, "/v2/library/fooimage/blobs/sha256:62d8908bee94c202b2d35224a221aaa2058318bfa9879fa541efaecba272331b"):
+				_, _ = w.Write(layerData)
+			default:
+				assert.FailNow(t, "unexpected path accessed: ", fmt.Sprintf("%s %s", r.URL.String(), tc.name))
+			}
+		}))
+		defer ts.Close()
+
+		c, err := client.NewClientWithOpts(client.WithHost(ts.URL))
+		assert.NoError(t, err)
+
+		// setup cache
+		s, f, err := setupCache()
+		defer func() {
+			_ = f.Close()
+			_ = os.RemoveAll(f.Name())
+		}()
+		assert.NoError(t, err, tc.name)
+
+		b := getTarZstdBuf(layerData)
+		if tc.cacheHit {
+			switch tc.garbageCache {
+			case true:
+				garbage, _ := ioutil.ReadFile("testdata/invalidgzvalidtar.tar.gz")
+				assert.NoError(t, s.Set(kvtypes.SetItemInput{
+					BucketName: "layertars",
+					Key:        string(inputDigest),
+					Value:      garbage,
+				}), tc.name)
+			default:
+				assert.NoError(t, s.Set(kvtypes.SetItemInput{
+					BucketName: "layertars",
+					Key:        string(inputDigest),
+					Value:      b,
+				}), tc.name)
+			}
+		}
+
+		de := Extractor{
+			Option: types.DockerOption{
+				AuthURL:  ts.URL,
+				NonSSL:   true,
+				SkipPing: true,
+				Timeout:  time.Second * 1000,
+			},
+			Client: c,
+			Cache:  s,
+		}
+
+		tsUrl := strings.TrimPrefix(ts.URL, "http://")
+		inputImage := registry.Image{
+			Domain: tsUrl,
+			Path:   "library/fooimage",
+			Tag:    "latest",
+		}
+
+		layerCh := make(chan layer)
+		errCh := make(chan error)
+		r, err := de.createRegistryClient(context.TODO(), inputImage.Domain)
+		go func() {
+			de.extractLayerWorker(inputDigest, r, context.TODO(), inputImage, errCh, layerCh, tc.requiredFiles)
+		}()
+
+		var errRecieved error
+		var layerReceived layer
+
+		select {
+		case errRecieved = <-errCh:
+			assert.FailNow(t, "unexpected error received, err: ", fmt.Sprintf("%s, %s", errRecieved, tc.name))
+		case layerReceived = <-layerCh:
+			assert.Equal(t, inputDigest, layerReceived.ID, tc.name)
+			assert.NotEmpty(t, layerReceived.Content, tc.name)
+			// TODO: Add tests to validate the content is sane
+		}
+
+		// check cache contents
+		var actualCacheContents []byte
+		found, err := s.Get(kvtypes.GetItemInput{
+			BucketName: "layertars",
+			Key:        string(inputDigest),
+			Value:      &actualCacheContents,
+		})
+
+		switch tc.cacheWB {
+		case true:
+			assert.True(t, found, tc.name)
+			assert.NoError(t, err, tc.name)
+
+			tbuf, err := extractTarContent(actualCacheContents)
```

Comment:
If a unit test depends on the application code, it can be a meaningless test. How about preparing a zstd file by manual and compare the result with it?

Reply:
Yeah I see your point. Although I was being over cautious here and trying to count all files the required files to be present in the cache by un-taring the cache contents and inspecting file by file. Maybe this is bit of an overkill. We could just manually store the golden file on disk and compare the two.

Reply:
I agree.